### PR TITLE
TRCL-3605 Static Typing: Subaccount Orders

### PR DESCRIPTION
**Pull request by ruixhuang**
Guarded by feature flag.  Cancellation doesn't work yet, and will address in follow-ups.

Screenshot_1721056416

### CODE REVIEW

File: src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt [hunk 1]
```diff
@@ -673,14 +673,19 @@ enum class OrderStatus(val rawValue: String) {
...
+        operator fun invoke(rawValue: String?): OrderStatus? {
             return if (rawValue == "BEST_EFFORT_OPENED") {
                 Pending
             } else {
                 OrderStatus.values().firstOrNull { it.rawValue == rawValue }
             }
         }
     }
+
+    val isFinalized: Boolean
```
Comment (aforaleka) [inline src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt:685]: 🙏

File: src/commonMain/kotlin/indexer/models/IndexerCompositeOrderObject.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import indexer.codegen.IndexerAPITimeInForce
+import indexer.codegen.IndexerIsoString
+import indexer.codegen.IndexerOrderSide
+import indexer.codegen.IndexerOrderType
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class IndexerCompositeOrderObject(
+
+    // Copied from IndexerOrderResponseObject
```
Comment (ruixhuang) [inline src/commonMain/kotlin/indexer/models/IndexerCompositeOrderObject.kt:13]: Websocket returns extra fields, so we can't use the codegen model directly

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt [hunk 3]
```diff
@@ -1264,7 +1293,7 @@ open class TradingStateMachine(
                 val modifiedFills = fills?.toIMutableMap() ?: mutableMapOf()
                 var subaccountFills = fills?.get(subaccountText)
                 if (staticTyping) {
-                    val newFills = internalState.wallet.account.subaccounts?.get(subaccountNumber)?.fills?.toIList()
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1267]: was this piece of internal state already released? 
just want to make sure we're isolating state changes

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt [hunk 4]
```diff
@@ -1264,7 +1293,7 @@ open class TradingStateMachine(
...
+                    val newFills = internalState.wallet.account.subaccounts[subaccountNumber]?.fills?.toIList()
```
Comment (yogurtandjam) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1296]: can we safely assume this if subaccount orders state is hidden behind a feature flag?

Comment (yogurtandjam) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:1296]: oh, this is behind the FF. sorry!

**Review by yogurtandjam — COMMENTED** [review]
can we add tests to make sure that this code behaves as expected for `staticTyping` true vs false?

**Comment by ruixhuang** [comment]
> can we add tests to make sure that this code behaves as expected for `staticTyping` true vs false?
> 
> doesn't have to be super heavy - it's probably fine if we just reuse an existing high level test and inject staticTyping true+ false into it and make sure all the outputs are the same.

Yes, you can set static_type = true when creating the state machine object during test.  However, there are some fields missing from Jared's market asset PR, so the tests would fail.  I will address this in the follow-up PR.

**Review by prashanDYDX — APPROVED** [review]

**Pull request merged**